cmd/hub/mcscontroller/controller.go

```diff
@@ -118,15 +121,205 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+			return ctrl.Result{}, err
+		}
 
+		err = exportSpec.IsCompatibleWith(primaryEs)
+		//3b:
+		if err != nil {
+			r.updateServiceExportConditions(ctx, &se, mcsv1a1.ServiceExportConflict, corev1.ConditionTrue, "", err.Error())
+			r.Client.Status().Update(ctx, &se)
+			//r.mcsClient.UpdateStatus(ctx, &se, metav1.UpdateOptions{})
+			err = r.deleteImport(ctx, &se)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		} else { //3c:
+			r.updateServiceExportConditions(ctx, &se, mcsv1a1.ServiceExportValid, corev1.ConditionTrue, "", "")
+			r.Client.Status().Update(ctx, &se)
+			//se, err := r.mcsClient.UpdateStatus(ctx, &se, metav1.UpdateOptions{})
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			r.ensureImportFor(ctx, &se)
+		}
+	}
 	return ctrl.Result{}, nil
 }
 
+// getPrimaryExportObject finds the Primary ExportSpec from ServiceExportList.
+// Retruns the ExportSpec object that determined as the primary one from
+// the given ServiceExportList according to the conflict resolution specification set in the
+// KEP, that is implemented in IsPreferredOver method of ExportSpec.
+func getPrimaryExportObject(exportList mcsv1a1.ServiceExportList) (*mcs.ExportSpec, error) {
+	var primary *mcs.ExportSpec
+	for i, se := range exportList.Items {
+		exportSpec := &mcs.ExportSpec{}
+		err := exportSpec.UnmarshalObjectMeta(&se.ObjectMeta)
+		if err != nil {
+			return nil, err
+		}
+		if i == 0 || exportSpec.IsPreferredOver(primary) {
+			primary = exportSpec
+		}
+	}
+	return primary, nil
+}
+
+// create or update the ServiceImport corresponding to the provided export.
+func (r *ServiceExportReconciler) ensureImportFor(ctx context.Context, se *mcsv1a1.ServiceExport) error {
+	es := &mcs.ExportSpec{}
+
+	err := es.UnmarshalObjectMeta(&se.ObjectMeta)
+	if err != nil {
+		return err
+	}
+	namespacedName := types.NamespacedName{
+		Namespace: es.Name,
+		Name:      es.Namespace,
+	}
+	log := r.Log.WithValues("service", namespacedName.Name)
+
+	si := mcsv1a1.ServiceImport{}
+	err = r.Client.Get(ctx, namespacedName, &si)
+	expected := mcsv1a1.ServiceImport{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: es.Namespace,
+			Name:      es.Name,
+		},
+		Spec: mcsv1a1.ServiceImportSpec{
+			Type:  es.Service.Type,
+			Ports: es.Service.Ports,
+		},
+	}
+	if apierrors.IsNotFound(err) {
+		err = r.Client.Create(ctx, &expected)
+		if err != nil {
+			log.Error(err, "unable to create the needed Service Import")
+			return err
+		}
+	} else if err == nil {
+		err = r.updateSi(ctx, &si, &expected, log)
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Error(err, "Failed to get the needed service import")
+		return err
+	}
+	return nil
+}
+
+func (r *ServiceExportReconciler) updateSi(ctx context.Context, si *mcsv1a1.ServiceImport, expected *mcsv1a1.ServiceImport, log logr.Logger) error {
+	//not sure it's really needed to check name and namespace - I got the si from the Get according to those fields
```

Comment:
please go over entire file and ensure there's a space between `//` and the first comment word